File: lab15/lab15.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>


void fill(int m, int n, double *(array[n]))
{
    for (int i = 0; i < m; i++)
    {
        for (int j = 0; j < n; j++)
        {
            array[i][j] = cos(sqrt(i)) - i + cos(j)/sqrt(1+j);
            // printf("%lf " , array[j][i]);
        }
        // printf("\n");
    }
}

void adT(int m, int n, double *(array[n]))
{
    int a;
    for(int j=0;j<m;j++)
    {
        for(int i=0;i<n;i++)
        {
            a=array[j][i];
            array[j][i]=array[i][j];
            array[i][j]=a;

            // printf("%lf " , array[j][i]);
        }
        // printf("\n");
    }

}

double find_min(int m, int n, double *(array[n]))
{
  double res = 100;
  for (int i = 0; i < m; i++)
      for (int j = 0; j < n; j++)
          if (fabs(array[i][j]) < sqrt(res))
              res = fabs(array[i][j]);
  return sqrt(res);

}

int main()
{
  int n, m;
  printf("Enter count of rows: ");
  scanf("%d", &m);
  printf("Enter count of columns: ");
  scanf("%d", &n);
  double **array = (double **)malloc(sizeof(double *) * m);
  if (!array)
  {
    printf("Memory allocation error!\n");
    exit(EXIT_FAILURE);
  }
  for (int i = 0; i < m; i++)
    array[i] = (double *)malloc(n * sizeof(double));

  fill(m, n, array);
  adT(m, n, array);
  double res = find_min(m, n, array);
  printf("%lf", res);
  return 0;
}

File: lab15/lab15.py
import math
print("Введите количество колон:")
m = int(input())
print("Введите количество строк--->")
n = int(input())

# m = 10
# n = 10

array = [[round(math.cos(math.sqrt(i)) - i + math.cos(j)/math.sqrt(1+j), 3) for i in range(m)]
         for j in range(n)]


# print(array)
for i in range(len(array)):
    for j in range(len(array[0])):
        print(array[i][j], end=' ')
    print()

trans_array = [[array[len(array) - 1 - j][len(array) - 1 - i] for i in range(len(array))]
             for j in range(len(array[0]))]

for i in range(len(array)):
    for j in range(len(array[0])):
        trans_array[i][j] = array[len(array) - j - 1][len(array[0]) - i - 1]
    # print()
print("ПОСЛЕ adT")

# print(trans_array)
for i in range(len(trans_array)):
    for j in range(len(trans_array[0])):
        print(trans_array[i][j], end=' ')
    print()


module_array = [[math.fabs(trans_array[i][j]) for i in range(len(trans_array))]
                for j in range(len(trans_array[0]))]


print("Модуль")
prost_array = []
for i in range(len(module_array)):
    for j in range(len(module_array[0])):
        print(module_array[i][j], end=' ')
    print()

umn_arr = []
print("*")

m = len(module_array)                                            # a: m × n
n = len(trans_array)                                            # b: n × k
k = len(trans_array[0])

umn_arr = [[None for __ in range(k)] for __ in range(m)]    # c: m × k

for i in range(m):
    for j in range(k):       
        umn_arr[i][j] = round(sum(module_array[i][kk] * trans_array[kk][j] for kk in range(n)), 2)
  
for i in range(len(umn_arr)):
    for j in range(len(umn_arr[0])):
        print(umn_arr[i][j], end=' ')
    print()

prost_array = []
for i in range(len(umn_arr)):
    for j in range(len(umn_arr[0])):
        prost_array.append(math.fabs(umn_arr[i][j]))
        # print(math.fabs(trans_array[i][j]), end=' ')
    # print()


print('САМЫЙ самый минимальный член: ', min(prost_array))

print('Контрольное значение: ', math.sqrt(min(prost_array)))
